### ISSUE TITLE
Slowness in tar_outdated()

### ISSUE THREAD

**Issue by wlandau** [wlandau/targets#108]
## Prework

* [x] I understand and agree to `targets`' [code of conduct](https://github.com/wlandau/targets/blob/master/CODE_OF_CONDUCT.md).
* [x] I understand and agree to `targets`' [contributing guidelines](https://github.com/wlandau/targets/blob/master/CONTRIBUTING.md).

## Description

The following profiling study reproduces the slowness somewhat. 

```{r}
library(proffer)
library(targets)
tar_script(
  tar_pipeline(
    tar_target(x, seq_len(1e3)),
    tar_target(y, x, pattern = map(x)),
    tar_target(z, y, pattern = map(y))
  )
)
tar_make(reporter = "silent")
px <- pprof(tar_outdated(reporter = "silent", callr_function = NULL))
```

Different systems show different bottlenecks. This flame graph was from a Mac.

![Screen Shot 2020-07-07 at 4 25 50 PM](https://user-images.githubusercontent.com/1580860/86838868-b000cf00-c06e-11ea-91af-44f1c4cb5ed2.png)

**Comment by wlandau** [comment]
Got rid of the most egregious bottlenecks. Not much we can do about inserting branches. Decrementing queue ranks is also just about as fast as we can get it in the current implementation.

![Screen Shot 2020-07-08 at 12 37 47 AM](https://user-images.githubusercontent.com/1580860/86876912-bebda500-c0b3-11ea-82a0-df430c7d8442.png)

Still about twice as slow as `drake`, but `outdated()` etc. always need to use the priority queue, and branch creation is more involved, so there's not much we can do. Maybe moving to a classical priority queue implemented in C or C++ we could get a max 15% speedup. We may revisit this issue later on, but it's no longer urgent.

**Comment by wlandau** [comment]
Sped up the queue updates a little more.

![Screen Shot 2020-07-08 at 1 02 21 AM](https://user-images.githubusercontent.com/1580860/86878462-eeba7780-c0b6-11ea-9895-8d2bf73882a1.png)

These are fast base R operations, and they might not show up as a bottleneck on other machines. Right now in this early stage, it does not seem worth going low level to speed up the queue, and we need the queue so behavior is consistent between `tar_outdated()` and the `tar_make*()` functions. That structure and rigor takes a little more time. The default `tar_outdated()` reporter should be a good stand-in for a progress bar so we don't leave users waiting and wondering.

**Comment by wlandau** [comment]
#111 may help.

**Comment by wlandau** [comment]
Fixed in #112. `tar_outdated()` is smarter than `r_outdated()` and pretty much just as fast, and there are no more obvious avoidable bottlenecks from the profiling data. That's good enough, especially for such a general issue.
